File: src/ee/ads/AdViewHelper.cpp
//
//  AdViewHelper.cpp
//  ee_x
//
//  Created by Zinge on 10/12/17.
//
//

#include "ee/ads/AdViewHelper.hpp"

namespace ee {
namespace ads {
using Self = AdViewHelper;

Self::AdViewHelper(const std::string& prefix, const std::string& adId) {
    prefix_ = prefix;
    adId_ = adId;
}

std::string Self::k__isLoaded() const {
    return prefix_ + "_isLoaded_" + adId_;
}

std::string Self::k__load() const {
    return prefix_ + "_load_" + adId_;
}

std::string Self::k__getPosition() const {
    return prefix_ + "_getPosition_" + adId_;
}

std::string Self::k__setPosition() const {
    return prefix_ + "_setPosition_" + adId_;
}

std::string Self::k__getSize() const {
    return prefix_ + "_getSize_" + adId_;
}

std::string Self::k__setSize() const {
    return prefix_ + "_setSize_" + adId_;
}

std::string Self::k__setVisible() const {
    return prefix_ + "_setVisible_" + adId_;
}
} // namespace ads
} // namespace ee

File: src/android/ads/src/main/java/com/ee/internal/AdViewHelper.kt
package com.ee.internal

import android.graphics.Point
import androidx.annotation.AnyThread
import com.ee.IAdView
import com.ee.IMessageBridge
import com.ee.Utils
import kotlinx.serialization.InternalSerializationApi
import kotlinx.serialization.Serializable

/**
 * Created by Zinge on 10/12/17.
 */
@InternalSerializationApi
class AdViewHelper(private val _bridge: IMessageBridge,
                   private val _view: IAdView,
                   private val _helper: MessageHelper) {
    @AnyThread
    fun registerHandlers() {
        _bridge.registerHandler(_helper.isLoaded) {
            Utils.toString(_view.isLoaded)
        }
        _bridge.registerHandler(_helper.load) {
            _view.load()
            ""
        }
        _bridge.registerHandler(_helper.getPosition) {
            @Serializable
            @Suppress("unused")
            class Response(
                val x: Int,
                val y: Int
            )

            val position = _view.position
            val response = Response(position.x, position.y)
            response.serialize()
        }
        _bridge.registerHandler(_helper.setPosition) { message ->
            @Serializable
            class Request(
                val x: Int,
                val y: Int
            )

            val request = deserialize<Request>(message)
            _view.position = Point(request.x, request.y)
            ""
        }
        _bridge.registerHandler(_helper.getSize) {
            @Serializable
            @Suppress("unused")
            class Response(
                val width: Int,
                val height: Int
            )

            val size = _view.size
            val response = Response(size.x, size.y)
            response.serialize()
        }
        _bridge.registerHandler(_helper.setSize) { message ->
            @Serializable
            class Request(
                val width: Int,
                val height: Int
            )

            val request = deserialize<Request>(message)
            _view.size = Point(request.width, request.height)
            ""
        }
        _bridge.registerHandler(_helper.isVisible) {
            Utils.toString(_view.isVisible)
        }
        _bridge.registerHandler(_helper.setVisible) { message ->
            _view.isVisible = Utils.toBoolean(message)
            ""
        }
    }

    @AnyThread
    fun deregisterHandlers() {
        _bridge.deregisterHandler(_helper.isLoaded)
        _bridge.deregisterHandler(_helper.load)
        _bridge.deregisterHandler(_helper.getPosition)
        _bridge.deregisterHandler(_helper.setPosition)
        _bridge.deregisterHandler(_helper.getSize)
        _bridge.deregisterHandler(_helper.setSize)
        _bridge.deregisterHandler(_helper.isVisible)
        _bridge.deregisterHandler(_helper.setVisible)
    }
}